File: generator-common/src/main/java/io/github/potjerodekool/openapi/common/generate/StandardOpenApiTypeUtils.java
package io.github.potjerodekool.openapi.common.generate;

import io.github.potjerodekool.codegen.model.type.TypeKind;
import io.github.potjerodekool.codegen.template.model.type.*;
import io.swagger.v3.oas.models.OpenAPI;
import io.swagger.v3.oas.models.media.*;

import java.util.List;
import java.util.Map;

public class StandardOpenApiTypeUtils implements OpenApiTypeUtils {
    @Override
    public TypeExpr createType(final OpenAPI openAPI, final Schema<?> schema,
                               final Map<String, Object> extensions,
                               final String packageName,
                               final String mediaType,
                               final Boolean isRequired) {
        return switch (schema) {
            case null -> new ClassOrInterfaceTypeExpr("java.lang.Object");
            case ArraySchema arraySchema -> createArrayType(openAPI, arraySchema, packageName, mediaType);
            case BinarySchema ignored -> createArrayType(new PrimitiveTypeExpr(TypeKind.BYTE));
            case BooleanSchema booleanSchema -> createBooleanType(booleanSchema, isRequired);
            case ByteArraySchema ignored -> throw new UnsupportedOperationException();
            case DateSchema ignored -> createDateType();
            case DateTimeSchema ignored -> createDateTimeType();
            case EmailSchema ignored -> createStringType(null);
            case FileSchema ignored -> throw new UnsupportedOperationException();
            case IntegerSchema integerSchema -> createIntegerType(integerSchema, isRequired);
            case JsonSchema ignored -> throw new UnsupportedOperationException();
            case MapSchema mapSchema -> createMapType(openAPI, mapSchema, packageName, mediaType, isRequired);
            case NumberSchema numberSchema -> createNumberType(numberSchema, isRequired);
            case ObjectSchema objectSchema -> {
                final var typeArg = ExtensionsHelper.getExtension(
                        objectSchema.getExtensions(),
                        Extensions.TYPE_ARG,
                        String.class
                );

                final ClassOrInterfaceTypeExpr type;

                if (typeArg == null) {
                    type = new ClassOrInterfaceTypeExpr().packageName("java.lang").simpleName("Object");
                } else {
                    type = new ClassOrInterfaceTypeExpr().simpleName(typeArg);
                }

                processExtensions(openAPI, type, packageName, extensions);
                yield type;
            }
            case PasswordSchema ignored -> createStringType(null);
            case StringSchema stringSchema -> createStringType(stringSchema);
            case UUIDSchema ignored -> createUuidType();
            default -> createTypeDefault(
                    openAPI,
                    schema,
                    extensions,
                    packageName,
                    mediaType,
                    isRequired);
        };
    }

    @Override
    public TypeExpr asNonNull(final TypeExpr typeExpr) {
        if (typeExpr instanceof PrimitiveTypeExpr) {
            return typeExpr;
        } else if (typeExpr instanceof ClassOrInterfaceTypeExpr classOrInterfaceTypeExpr) {
            final var name = classOrInterfaceTypeExpr.getName();

            return switch (name) {
                case "java.lang.Boolean":
                    yield new ClassOrInterfaceTypeExpr("java.lang.Boolean");
                case "java.lang.Integer":
                    yield new ClassOrInterfaceTypeExpr("java.lang.Integer");
                case "java.lang.Long":
                    yield new ClassOrInterfaceTypeExpr("java.lang.Long");
                case "java.lang.Float":
                    yield new ClassOrInterfaceTypeExpr("java.lang.Float");
                case "java.lang.Double":
                    yield new ClassOrInterfaceTypeExpr("java.lang.Double");
                default:
                    yield typeExpr;
            };
        }

        throw new UnsupportedOperationException();
    }

    @Override
    public TypeExpr createMultipartTypeExpression(final OpenAPI api) {
        throw new UnsupportedOperationException();
    }

    @Override
    public TypeExpr resolveImplementationType(final OpenAPI openAPI, final TypeExpr type) {
        final TypeExpr implementationType;

        if (type instanceof WildCardTypeExpr wildcardType) {
            implementationType = (TypeExpr) wildcardType.getExpr();
        } else {
            implementationType = type;
        }
        return implementationType;
    }

    @Override
    public TypeExpr createNumberType(final NumberSchema numberSchema, final Boolean isRequired) {
        final var isNullable = Boolean.TRUE.equals(numberSchema.getNullable());

        if ("double".equals(numberSchema.getFormat())) {
            return isNullable || Boolean.FALSE.equals(isRequired)
                    ? new ClassOrInterfaceTypeExpr("java.lang.Double")
                    : new PrimitiveTypeExpr(TypeKind.DOUBLE);
        } else if ("float".equals(numberSchema.getFormat())) {
            return isNullable || Boolean.FALSE.equals(isRequired)
                    ? new ClassOrInterfaceTypeExpr("java.lang.Float")
                    : new PrimitiveTypeExpr(TypeKind.FLOAT);
        } else {
            return new ClassOrInterfaceTypeExpr("java.math.BigDecimal");
        }
    }

    @Override
    public TypeExpr createStringType(final StringSchema schema) {
        if (schema != null && schema.getEnum() != null) {
            return new ClassOrInterfaceTypeExpr().packageName("java.lang").simpleName("Enum");
        }

        return new ClassOrInterfaceTypeExpr().packageName("java.lang").simpleName("String");
    }

    @Override
    public TypeExpr createDateType() {
        return new ClassOrInterfaceTypeExpr("java.time.LocalDate");
    }

    @Override
    public TypeExpr createDateTimeType() {
        return new ClassOrInterfaceTypeExpr("java.time.OffsetDateTime");
    }

    @Override
    public TypeExpr createBooleanType(final BooleanSchema booleanSchema,
                                      final Boolean isRequired) {
        final var isNullable = Boolean.TRUE.equals(booleanSchema.getNullable());
        return isNullable || Boolean.FALSE.equals(isRequired)
                ? new ClassOrInterfaceTypeExpr("java.lang.Boolean")
                : new PrimitiveTypeExpr(TypeKind.BOOLEAN);
    }

    @Override
    public TypeExpr createUuidType() {
        return new ClassOrInterfaceTypeExpr("java.util.UUID");
    }

    @Override
    public ClassOrInterfaceTypeExpr createMapType(final OpenAPI openAPI,
                                                  final MapSchema mapSchema,
                                                  final String packageName,
                                                  final String mediaType,
                                                  final Boolean isRequired) {
        final var keyType = new ClassOrInterfaceTypeExpr("java.lang.String");
        final var valueType = createType(
                openAPI,
                (Schema<?>) mapSchema.getAdditionalProperties(),
                null,
                packageName,
                mediaType,
                isRequired
        );

        return new ClassOrInterfaceTypeExpr("java.util.Map")
                .typeArguments(
                        keyType,
                        valueType
                );
    }

    @Override
    public TypeExpr createArrayType(final OpenAPI openAPI,
                                    final ArraySchema arraySchema,
                                    final String packageName,
                                    final String mediaType) {
        final var componentType = createType(
                openAPI,
                arraySchema.getItems(),
                null,
                packageName,
                mediaType,
                false
        );

        if (componentType instanceof PrimitiveTypeExpr primitiveType) {
            return createArrayType(primitiveType);
        } else {
            final var className = Boolean.TRUE.equals(arraySchema.getItems().getUniqueItems())
                    ? "java.util.Set"
                    : "java.util.List";

            return new ClassOrInterfaceTypeExpr(className)
                    .typeArgument(componentType);
        }
    }

    private ArrayTypeExpr createArrayType(final PrimitiveTypeExpr primitiveType) {
        return new ArrayTypeExpr(primitiveType);
    }

    private TypeExpr createIntegerType(final IntegerSchema integerSchema,
                                       final Boolean isRequired) {
        final var isNullable = Boolean.TRUE.equals(integerSchema.getNullable());

        if ("int64".equals(integerSchema.getFormat())) {
            return isNullable || Boolean.FALSE.equals(isRequired)
                    ? new ClassOrInterfaceTypeExpr("java.lang.Long")
                    : new PrimitiveTypeExpr(TypeKind.LONG);
        } else {
            return isNullable || Boolean.FALSE.equals(isRequired)
                    ? new ClassOrInterfaceTypeExpr("java.lang.Integer")
                    : new PrimitiveTypeExpr(TypeKind.INT);
        }
    }

    private void processExtensions(final OpenAPI openAPI,
                                   final ClassOrInterfaceTypeExpr type,
                                   final String packageName,
                                   final Map<String, Object> extensions) {
        final var typeArgs = ExtensionsHelper.getExtension(extensions, Extensions.TYPE_ARGS, List.class);

        if (typeArgs != null) {
            for (final Object typeArg : typeArgs) {
                if (typeArg instanceof String className) {
                    final var typeArgType = new ClassOrInterfaceTypeExpr(className);
                    type.typeArgument(typeArgType);
                } else if (typeArg instanceof Map<?, ?> map) {
                    final var ref = (String) map.get("$ref");

                    if (ref != null) {
                        final var resolvedSchema = SchemaResolver.resolve(openAPI, ref);

                        final var typeArgType = createType(
                                openAPI,
                                resolvedSchema.schema(),
                                null,
                                packageName,
                                null,
                                true
                        );

                        if (typeArgType instanceof ClassOrInterfaceTypeExpr classOrInterfaceTypeExpr) {
                            classOrInterfaceTypeExpr.name(packageName + "." + resolvedSchema.name());
                        }

                        type.typeArgument(typeArgType);
                    }
                }
            }
        }
    }

    private TypeExpr createTypeDefault(final OpenAPI openAPI,
                                       final Schema<?> schema,
                                       final Map<String, Object> extensions,
                                       final String packageName,
                                       final String mediaType,
                                       final Boolean isRequired) {
        final var resolved = SchemaResolver.resolve(openAPI, schema);
        final var resolvedSchema = resolved.schema();

        return switch (resolvedSchema) {
            case null -> throw new NullPointerException("Resolved schema is null");
            case ObjectSchema ignored -> createObjectOrComposedType(openAPI, resolved, extensions, packageName);
            case ComposedSchema ignored -> createObjectOrComposedType(openAPI, resolved, extensions, packageName);
            default -> {
                if (resolvedSchema.getClass() == Schema.class) {
                    throw new UnsupportedOperationException();
                } else {
                    yield createType(
                            openAPI,
                            resolved.schema(),
                            extensions,
                            packageName,
                            mediaType,
                            isRequired
                    );
                }
            }
        };
    }

    private TypeExpr createObjectOrComposedType(final OpenAPI openAPI,
                                                final ResolvedSchemaResult resolved,
                                                final Map<String, Object> extensions,
                                                final String packageName) {
        final var name = resolved.name();
        if (name != null) {
            final var type = new ClassOrInterfaceTypeExpr(packageName + "." + resolved.name());
            processExtensions(openAPI, type, packageName, extensions);
            return type;
        } else {
            return new ClassOrInterfaceTypeExpr("java.lang.Object");
        }
    }
}

File: generator-common/src/main/kotlin/io/github/potjerodekool/openapi/common/generate/StandardOpenApiTypeUtils.kt
package io.github.potjerodekool.openapi.common.generate

import io.github.potjerodekool.codegen.model.type.TypeKind
import io.github.potjerodekool.codegen.template.model.type.*
import io.github.potjerodekool.openapi.common.SchemaType
import io.github.potjerodekool.openapi.common.SchemaTypeResolver
import io.swagger.v3.oas.models.OpenAPI
import io.swagger.v3.oas.models.media.*

open class StandardOpenApiTypeUtils : OpenApiTypeUtils {

    override fun createType(
        openAPI: OpenAPI,
        schema: Schema<*>?,
        extensions: MutableMap<String, Any?>?,
        packageName: String,
        mediaType: ContentType?,
        isRequired: Boolean?
    ): TypeExpr {
        if (schema == null) {
            return ClassOrInterfaceTypeExpr("java.lang.Object")
        }

        val schemaType = SchemaTypeResolver.resolveSchemaType(schema)

        return when (schemaType) {
            SchemaType.ARRAY -> createArrayType(openAPI, schema, packageName, mediaType)
            SchemaType.BINARY -> createMultipartTypeExpression(openAPI)
            SchemaType.BOOLEAN -> createBooleanType(schema, isRequired)
            SchemaType.DATE -> createDateType()
            SchemaType.DATE_TIME -> createDateTimeType()
            SchemaType.EMAIL -> createStringType(null)
            SchemaType.INT32, SchemaType.INT64 -> createIntegerType(schema, isRequired)
            SchemaType.MAP -> createMapType(openAPI, schema, packageName, mediaType, isRequired)
            SchemaType.FLOAT, SchemaType.DOUBLE -> createNumberType(schema, isRequired)
            SchemaType.OBJECT -> {
                val typeArg = ExtensionsHelper.getExtension(
                    schema.extensions,
                    Extensions.TYPE_ARG,
                    String::class.java
                )

                val type = if (typeArg == null) {
                    ClassOrInterfaceTypeExpr("java.lang.Object")
                } else {
                    ClassOrInterfaceTypeExpr(typeArg)
                }

                processExtensions(openAPI, type, packageName, extensions)
                type
            }
            SchemaType.PASSWORD -> createStringType(null)
            SchemaType.STRING -> createStringType(schema)
            SchemaType.UUID -> createUuidType()
            else -> {
                when (schema) {
                    is ByteArraySchema -> throw UnsupportedOperationException()
                    is FileSchema -> throw UnsupportedOperationException()
                    else -> createTypeDefault(
                        openAPI,
                        schema,
                        extensions,
                        packageName,
                        mediaType,
                        isRequired
                    )
                }
            }
        }
    }

    override fun asNonNull(typeExpr: TypeExpr?): TypeExpr {
        if (typeExpr is PrimitiveTypeExpr) {
            return typeExpr
        } else if (typeExpr is ClassOrInterfaceTypeExpr) {
            val name = typeExpr.name
            return when (name) {
                "java.lang.Boolean" -> ClassOrInterfaceTypeExpr("java.lang.Boolean")
                "java.lang.Integer" -> ClassOrInterfaceTypeExpr("java.lang.Integer")
                "java.lang.Long" -> ClassOrInterfaceTypeExpr("java.lang.Long")
                "java.lang.Float" -> ClassOrInterfaceTypeExpr("java.lang.Float")
                "java.lang.Double" -> ClassOrInterfaceTypeExpr("java.lang.Double")
                else -> typeExpr
            }
        }

        throw UnsupportedOperationException()
    }

    private fun asNullable(typeExpr: TypeExpr): TypeExpr {
        return when (typeExpr.typeKind) {
            TypeKind.BOOLEAN -> ClassOrInterfaceTypeExpr("java.lang.Boolean")
            TypeKind.INT -> ClassOrInterfaceTypeExpr("java.lang.Integer")
            TypeKind.LONG -> ClassOrInterfaceTypeExpr("java.lang.Long")
            TypeKind.FLOAT -> ClassOrInterfaceTypeExpr("java.lang.Float")
            TypeKind.DOUBLE -> ClassOrInterfaceTypeExpr("java.lang.Double")
            TypeKind.BYTE -> ClassOrInterfaceTypeExpr("java.lang.Byte")
            TypeKind.SHORT -> ClassOrInterfaceTypeExpr("java.lang.Short")
            TypeKind.CHAR -> ClassOrInterfaceTypeExpr("java.lang.Character")
            else -> typeExpr
        }
    }

    override fun createMultipartTypeExpression(api: OpenAPI): TypeExpr {
        throw UnsupportedOperationException()
    }

    override fun resolveImplementationType(openAPI: OpenAPI, type: TypeExpr?): TypeExpr? {
        return if (type is WildCardTypeExpr) {
            type.expr as TypeExpr
        } else {
            type
        }
    }

    override fun createNumberType(numberSchema: Schema<*>, isRequired: Boolean?): TypeExpr {
        val isNullable = true == numberSchema.nullable

        return if ("double" == numberSchema.format) {
            if (isNullable || false == isRequired) ClassOrInterfaceTypeExpr("java.lang.Double")
            else PrimitiveTypeExpr(TypeKind.DOUBLE)
        } else if ("float" == numberSchema.format) {
            if (isNullable || false == isRequired) ClassOrInterfaceTypeExpr("java.lang.Float")
            else PrimitiveTypeExpr(TypeKind.FLOAT)
        } else {
            ClassOrInterfaceTypeExpr("java.math.BigDecimal")
        }
    }

    override fun createStringType(schema: Schema<*>?): TypeExpr {
        return if (schema?.enum != null) {
            ClassOrInterfaceTypeExpr().packageName("java.lang").simpleName("Enum")
        } else {
            ClassOrInterfaceTypeExpr().packageName("java.lang").simpleName("String")
        }
    }

    override fun createDateType(): TypeExpr {
        return ClassOrInterfaceTypeExpr("java.time.LocalDate")
    }

    override fun createDateTimeType(): TypeExpr {
        return ClassOrInterfaceTypeExpr("java.time.OffsetDateTime")
    }

    override fun createBooleanType(booleanSchema: Schema<*>, isRequired: Boolean?): TypeExpr {
        val isNullable = true == booleanSchema.nullable
        return if (isNullable || false == isRequired) ClassOrInterfaceTypeExpr("java.lang.Boolean")
        else PrimitiveTypeExpr(TypeKind.BOOLEAN)
    }

    override fun createUuidType(): TypeExpr {
        return ClassOrInterfaceTypeExpr("java.util.UUID")
    }

    override fun createMapType(
        openAPI: OpenAPI,
        mapSchema: Schema<*>,
        packageName: String,
        mediaType: ContentType?,
        isRequired: Boolean?
    ): ClassOrInterfaceTypeExpr {
        val keyType = ClassOrInterfaceTypeExpr("java.lang.String")
        val valueType = asNullable(
            createType(
                openAPI,
                mapSchema.additionalProperties as Schema<*>,
                mutableMapOf(),
                packageName,
                mediaType,
                isRequired
            )
        )

        return ClassOrInterfaceTypeExpr("java.util.Map")
            .typeArguments(
                keyType,
                valueType
            )
    }

    override fun createArrayType(
        openAPI: OpenAPI,
        arraySchema: Schema<*>,
        packageName: String,
        mediaType: ContentType?
    ): TypeExpr {
        val componentType = createType(
            openAPI,
            arraySchema.items,
            mutableMapOf(),
            packageName,
            mediaType,
            false
        )

        if (componentType is PrimitiveTypeExpr) {
            return createArrayType(componentType)
        } else {
            val className =
                if (true == arraySchema.items.uniqueItems) "java.util.Set"
                else "java.util.List"

            return ClassOrInterfaceTypeExpr(className)
                .typeArgument(componentType)
        }
    }

    private fun createArrayType(primitiveType: PrimitiveTypeExpr): ArrayTypeExpr {
        return ArrayTypeExpr(primitiveType)
    }

    override fun isCollectionType(typeExpr: TypeExpr?): Boolean {
        if (typeExpr is ClassOrInterfaceTypeExpr) {
            return "java.util.Set" == typeExpr.name
                    || "java.util.List" == typeExpr.name
        }
        return false
    }

    private fun createIntegerType(
        integerSchema: Schema<*>,
        isRequired: Boolean?
    ): TypeExpr {
        val isNullable = true == integerSchema.nullable

        return if ("int64" == integerSchema.format) {
            if (isNullable || false == isRequired) ClassOrInterfaceTypeExpr("java.lang.Long")
            else PrimitiveTypeExpr(TypeKind.LONG)
        } else {
            if (isNullable || false == isRequired) ClassOrInterfaceTypeExpr("java.lang.Integer")
            else PrimitiveTypeExpr(TypeKind.INT)
        }
    }

    private fun processExtensions(
        openAPI: OpenAPI, type: ClassOrInterfaceTypeExpr,
        packageName: String,
        extensions: MutableMap<String, Any?>?
    ) {
        val typeArgs = ExtensionsHelper.getExtension(extensions, Extensions.TYPE_ARGS, List::class.java)

        if (typeArgs != null) {
            typeArgs.forEach { typeArg ->
                if (typeArg is String) {
                    val typeArgType = ClassOrInterfaceTypeExpr(typeArg)
                    type.typeArgument(typeArgType)
                } else if (typeArg is Map<*, *>) {
                    val ref = typeArg["\$ref"] as String?

                    if (ref != null) {
                        val resolvedSchema = SchemaResolver.resolve(openAPI, ref)

                        val typeArgType = createType(
                            openAPI,
                            resolvedSchema.schema(),
                            mutableMapOf(),
                            packageName,
                            null,
                            true
                        )

                        if (typeArgType is ClassOrInterfaceTypeExpr) {
                            typeArgType.name(packageName + "." + resolvedSchema.name())
                        }

                        type.typeArgument(typeArgType)
                    }
                }
            }
        }
    }

    private fun createTypeDefault(
        openAPI: OpenAPI,
        schema: Schema<*>,
        extensions: MutableMap<String, Any?>?,
        packageName: String,
        mediaType: ContentType?,
        isRequired: Boolean?
    ): TypeExpr {
        val resolved = SchemaResolver.resolve(openAPI, schema)
        return when (val resolvedSchema = resolved.schema()) {
            null -> throw NullPointerException("Resolved schema is null")
            is ObjectSchema -> createObjectOrComposedType(openAPI, resolved, extensions, packageName)
            is ComposedSchema -> createObjectOrComposedType(openAPI, resolved, extensions, packageName)
            else -> {
                if (resolvedSchema.javaClass == Schema::class.java) {
                    throw UnsupportedOperationException()
                } else {
                    return createType(
                        openAPI,
                        resolved.schema(),
                        extensions,
                        packageName,
                        mediaType,
                        isRequired
                    )
                }
            }
        }
    }

    private fun createObjectOrComposedType(
        openAPI: OpenAPI,
        resolved: ResolvedSchemaResult,
        extensions: MutableMap<String, Any?>?,
        packageName: String
    ): TypeExpr {
        val name = resolved.name()
        if (name != null) {
            val type = ClassOrInterfaceTypeExpr (packageName + "." + resolved.name())
            processExtensions(openAPI, type, packageName, extensions)
            return type
        } else {
            return ClassOrInterfaceTypeExpr ("java.lang.Object")
        }
    }
}